model/jpa/src/main/resources/META-INF/jpa-changelog-24.0.0.xml

```diff
@@ -50,4 +50,149 @@
+            FROM USER_CONSENT
+            GROUP BY CLIENT_ID, USER_ID
+            HAVING COUNT(*) > 1
+            ) max_dates ON uc.CLIENT_ID = max_dates.CLIENT_ID
+            AND uc.USER_ID = max_dates.USER_ID
+            AND uc.LAST_UPDATED_DATE = max_dates.MAX_UPDATED_DATE
+            );
+
+            CREATE TEMPORARY TABLE TEMP_USER_CONSENT_IDS AS
+            SELECT uc.ID
+            FROM USER_CONSENT uc
+            INNER JOIN (
+            SELECT CLIENT_ID, USER_ID, MAX(LAST_UPDATED_DATE) AS MAX_UPDATED_DATE
+            FROM USER_CONSENT
+            GROUP BY CLIENT_ID, USER_ID
+            HAVING COUNT(*) > 1
+            ) max_dates ON uc.CLIENT_ID = max_dates.CLIENT_ID
+            AND uc.USER_ID = max_dates.USER_ID
+            AND uc.LAST_UPDATED_DATE = max_dates.MAX_UPDATED_DATE;
+
+            DELETE FROM USER_CONSENT
+            WHERE ID IN (SELECT ID FROM TEMP_USER_CONSENT_IDS);
+
+            DROP TEMPORARY TABLE IF EXISTS TEMP_USER_CONSENT_IDS;
+
+            DELETE FROM USER_CONSENT_CLIENT_SCOPE
+            WHERE USER_CONSENT_ID IN (
+            SELECT uc.ID
+            FROM USER_CONSENT uc
+            INNER JOIN (
+            SELECT CLIENT_STORAGE_PROVIDER, EXTERNAL_CLIENT_ID, USER_ID, MAX(LAST_UPDATED_DATE) AS MAX_UPDATED_DATE
+            FROM USER_CONSENT
+            GROUP BY CLIENT_STORAGE_PROVIDER, EXTERNAL_CLIENT_ID, USER_ID
+            HAVING COUNT(*) > 1
+            ) max_dates ON uc.CLIENT_STORAGE_PROVIDER = max_dates.CLIENT_STORAGE_PROVIDER
+            AND uc.EXTERNAL_CLIENT_ID = max_dates.EXTERNAL_CLIENT_ID
+            AND uc.USER_ID = max_dates.USER_ID
+            AND uc.LAST_UPDATED_DATE = max_dates.MAX_UPDATED_DATE
+            );
+
+            CREATE TEMPORARY TABLE TEMP_USER_CONSENT_IDS2 AS
+            SELECT uc.ID
+            FROM USER_CONSENT uc
+            INNER JOIN (
+            SELECT CLIENT_STORAGE_PROVIDER, EXTERNAL_CLIENT_ID, USER_ID, MAX(LAST_UPDATED_DATE) AS MAX_UPDATED_DATE
+            FROM USER_CONSENT
+            GROUP BY CLIENT_STORAGE_PROVIDER, EXTERNAL_CLIENT_ID, USER_ID
+            HAVING COUNT(*) > 1
+            ) max_dates ON uc.CLIENT_STORAGE_PROVIDER = max_dates.CLIENT_STORAGE_PROVIDER
+            AND uc.EXTERNAL_CLIENT_ID = max_dates.EXTERNAL_CLIENT_ID
+            AND uc.USER_ID = max_dates.USER_ID
+            AND uc.LAST_UPDATED_DATE = max_dates.MAX_UPDATED_DATE;
+
+            DELETE FROM USER_CONSENT
+            WHERE ID IN (SELECT ID FROM TEMP_USER_CONSENT_IDS2);
+
+            DROP TEMPORARY TABLE IF EXISTS TEMP_USER_CONSENT_IDS2;
+        </sql>
+        <dropUniqueConstraint tableName="USER_CONSENT" constraintName="UK_JKUWUVD56ONTGSUHOGM8UEWRT"/>
+        <addUniqueConstraint columnNames="CLIENT_ID, USER_ID" constraintName="UK_LOCAL_CONSENT" tableName="USER_CONSENT"/>
+        <addUniqueConstraint columnNames="CLIENT_STORAGE_PROVIDER, EXTERNAL_CLIENT_ID, USER_ID" constraintName="UK_EXTERNAL_CONSENT" tableName="USER_CONSENT"/>
+    </changeSet>
+
```

Comment:
In Keycloak we have avoided the explicit SQL statements in the changesets, and instead we use custom changes like the one in [this PR](https://github.com/keycloak/keycloak/commit/e49dfe534e979db7f401021b7e2bb7d8c4c7cd73#diff-79a74c9de3a029d446c502274e5709f364a0787cc507183cfeed84d86b466cf6) -  for consistency I would use the same approach here.

Reply:
> In Keycloak we have avoided the explicit SQL statements in the changesets, and instead we use custom changes like the one in [this PR](https://github.com/keycloak/keycloak/commit/e49dfe534e979db7f401021b7e2bb7d8c4c7cd73#diff-79a74c9de3a029d446c502274e5709f364a0787cc507183cfeed84d86b466cf6) - for consistency I would use the same approach here.

Ok, I will check it and will make appropriate changes if possible.

Reply:
I have migrate SQL statements to Java class extends CustomKeycloakTask.